### PR TITLE
[Superseded] Add Bake Mask to GPUParticlesCollisionHeightField3D, Collision Mask to GPUParticles3D and Collision Layer to GPUParticlesCollision3D & GPUParticlesAttractor3D.

### LINKED ISSUE #61014: Vulkan: GPUParticles3D attractor and collision Cull Mask doesn't have any effect
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0.alpha8

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 510.60.02)

### Issue description

GPUParticles3D attractor and collision Cull Mask doesn't have any effect. I noticed this while working on https://github.com/godotengine/godot/pull/60992.

The same issue occurs with GPUParticles3D collision nodes.

*Left: GPUParticles3D on layer 1, GPUParticlesAttractorBox3D with all cull mask layers enabled (default)*
*Middle: GPUParticles3D on layer 2, GPUParticlesAttractorBox3D with all cull mask layers disabled*
*Right: GPUParticles3D on layer 2, GPUParticlesAttractorBox3D with only the cull mask layer 2 enabled*

### Steps to reproduce

- Add a GPUParticles3D node, a draw pass and a process material to it. Set the emission shape to a sphere to you can see individual particles separated from each other.
- Add a GPUParticlesAttractor3DBox node.
- Enable only the second visual layer on the GPUParticles3D node.
- Set the cull mask to exclude all layers on the GPUParticlesAttractorBox3D node.
- To reproduce the same issue with collision, add GPUParticlesCollisionBox3D nodes and enable collision on the particles' Process Material.

### Minimal reproduction project

[test_particle_attractors_cull_mask.zip](https://github.com/godotengine/godot/files/8691548/test_particle_attractors_cull_mask.zip)

### PR DESCRIPTION
**Pull request by yahiaetman**
This pull request mainly adds a `bake_mask` field for `GPUParticlesCollisionHeightField3D` that functions similarly to the `bake_mask` in `GPUParticlesCollisionSDF3D`. It is used to filter which meshes are rendered in the height field texture.

However, my goal was to have separate collisions that reacts to different meshes and affects separate `GPUParticles3D` nodes. The problem was that the field `cull_mask` in `GPUParticlesCollision3D` did not work as noted in this issue https://github.com/godotengine/godot/issues/61014#issue-1235751692

Looking at the code, the field `cull_mask` was not used anywhere, so I replaced it with `collision_layer` in both `GPUParticlesCollision3D` and `GPUParticlesAttractor3D` which uses a new set of layers under the name `"layer_names/3d_particle_collision"`. For backward compatibility, the `cull_mask` property is kept as a deprecated alias to the  `collision_layer` property. In the `GPUParticles3D` class, a corresponding field `collision_mask` was added. I argue that this would be easier for the users than using the 3D render layers for both the rendering culling masks and the 3D particle collision masks.

If any changes are needed or if you prefer that I split the `GPUParticlesCollisionHeightField3D.bake_mask` code to a separate pull request, please let me know.

And thank you for your amazing work on this wonderful game engine.

- *Production edit: This closes https://github.com/godotengine/godot/issues/61014.*

### CODE REVIEW

File: scene/3d/gpu_particles_collision_3d.cpp [hunk 1]
```diff
@@ -851,8 +918,15 @@ real_t GPUParticlesAttractor3D::get_directionality() const {
...
 
 void GPUParticlesAttractor3D::_bind_methods() {
-	ClassDB::bind_method(D_METHOD("set_cull_mask", "mask"), &GPUParticlesAttractor3D::set_cull_mask);
-	ClassDB::bind_method(D_METHOD("get_cull_mask"), &GPUParticlesAttractor3D::get_cull_mask);
+	ClassDB::bind_method(D_METHOD("set_collision_layer", "layer"), &GPUParticlesAttractor3D::set_collision_layer);
+	ClassDB::bind_method(D_METHOD("get_collision_layer"), &GPUParticlesAttractor3D::get_collision_layer);
+
+#ifndef DISABLE_DEPRECATED
+
+	ClassDB::bind_compatibility_method(D_METHOD("set_cull_mask", "mask"), &GPUParticlesAttractor3D::set_cull_mask);
```
Comment (AThousandShips) [inline scene/3d/gpu_particles_collision_3d.cpp:926]: You don't use `bind_compatibility_method` for this, just use the normal method

File: scene/3d/gpu_particles_collision_3d.cpp [hunk 2]
```diff
@@ -866,7 +940,8 @@ void GPUParticlesAttractor3D::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "strength", PROPERTY_HINT_RANGE, "-128,128,0.01,or_greater,or_less"), "set_strength", "get_strength");
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "attenuation", PROPERTY_HINT_EXP_EASING, "0,8,0.01"), "set_attenuation", "get_attenuation");
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "directionality", PROPERTY_HINT_RANGE, "0,1,0.01"), "set_directionality", "get_directionality");
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_RENDER), "set_cull_mask", "get_cull_mask");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "collision_layer", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION), "set_collision_layer", "get_collision_layer");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION), "set_cull_mask", "get_cull_mask");
```
Comment (AThousandShips) [inline scene/3d/gpu_particles_collision_3d.cpp:944]: The properties should also have `#ifndef DISABLE_DEPRECATED`

File: scene/3d/gpu_particles_collision_3d.cpp [hunk 3]
```diff
@@ -866,7 +943,13 @@ void GPUParticlesAttractor3D::_bind_methods() {
...
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "attenuation", PROPERTY_HINT_EXP_EASING, "0,8,0.01"), "set_attenuation", "get_attenuation");
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "directionality", PROPERTY_HINT_RANGE, "0,1,0.01"), "set_directionality", "get_directionality");
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_RENDER), "set_cull_mask", "get_cull_mask");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "collision_layer", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION), "set_collision_layer", "get_collision_layer");
+
+#ifndef DISABLE_DEPRECATED
+
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION), "set_cull_mask", "get_cull_mask");
+
+#endif //DISABLE_DEPRECATED
```
Comment (AThousandShips) [inline scene/3d/gpu_particles_collision_3d.cpp:952]:
```suggestion
#ifndef DISABLE_DEPRECATED
	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION, "", PROPERTY_USAGE_NONE), "set_cull_mask", "get_cull_mask");
#endif // DISABLE_DEPRECATED
```
This should make sure they are neither stored nor visible in the editor

File: scene/3d/gpu_particles_collision_3d.cpp [hunk 4]
```diff
@@ -851,9 +921,16 @@ real_t GPUParticlesAttractor3D::get_directionality() const {
...
 void GPUParticlesAttractor3D::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_collision_layer", "layer"), &GPUParticlesAttractor3D::set_collision_layer);
+	ClassDB::bind_method(D_METHOD("get_collision_layer"), &GPUParticlesAttractor3D::get_collision_layer);
+
+#ifndef DISABLE_DEPRECATED
+
 	ClassDB::bind_method(D_METHOD("set_cull_mask", "mask"), &GPUParticlesAttractor3D::set_cull_mask);
 	ClassDB::bind_method(D_METHOD("get_cull_mask"), &GPUParticlesAttractor3D::get_cull_mask);
 
+#endif //DISABLE_DEPRECATED
```
Comment (AThousandShips) [inline scene/3d/gpu_particles_collision_3d.cpp:932]:
```suggestion
#ifndef DISABLE_DEPRECATED
	ClassDB::bind_method(D_METHOD("set_cull_mask", "mask"), &GPUParticlesAttractor3D::set_cull_mask);
	ClassDB::bind_method(D_METHOD("get_cull_mask"), &GPUParticlesAttractor3D::get_cull_mask);
#endif // DISABLE_DEPRECATED
```

File: scene/3d/gpu_particles_collision_3d.cpp [hunk 5]
```diff
@@ -813,14 +872,25 @@ GPUParticlesCollisionHeightField3D::~GPUParticlesCollisionHeightField3D() {
...
-	RS::get_singleton()->particles_collision_set_cull_mask(collision, p_cull_mask);
+	WARN_DEPRECATED_MSG(R"(The "cull_mask" property is deprecated, use "collision_layer" instead.)");
+	set_collision_layer(p_cull_mask);
 }
 
 uint32_t GPUParticlesAttractor3D::get_cull_mask() const {
-	return cull_mask;
+	return collision_layer;
 }
+#endif //DISABLE_DEPRECATED
```
Comment (AThousandShips) [inline scene/3d/gpu_particles_collision_3d.cpp:893]:
```suggestion
#endif // DISABLE_DEPRECATED
```

File: scene/3d/gpu_particles_collision_3d.cpp [hunk 6]
```diff
@@ -35,20 +35,46 @@
...
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_RENDER), "set_cull_mask", "get_cull_mask");
+#endif //DISABLE_DEPRECATED
+
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "collision_layer", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION), "set_collision_layer", "get_collision_layer");
+
+#ifndef DISABLE_DEPRECATED
+
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION), "set_cull_mask", "get_cull_mask");
+
+#endif //DISABLE_DEPRECATED
```
Comment (AThousandShips) [inline scene/3d/gpu_particles_collision_3d.cpp:77]:
```suggestion
#ifndef DISABLE_DEPRECATED
	ADD_PROPERTY(PropertyInfo(Variant::INT, "cull_mask", PROPERTY_HINT_LAYERS_3D_PARTICLE_COLLISION, "", PROPERTY_USAGE_NONE), "set_cull_mask", "get_cull_mask");
#endif // DISABLE_DEPRECATED
```

**Comment by yahiaetman** [comment]
I changed it temporarily to a draft till I get the checks to pass, then I will turn it to a pull request again as soon as possible.

File: scene/3d/gpu_particles_collision_3d.cpp (hunk 2 above)
Comment (yahiaetman) [reply scene/3d/gpu_particles_collision_3d.cpp:950]: Currently, I have a redundant field in the editor for `cull_mask` which shows exactly the same thing as `collision_layer`. Is there a way to hide it in the editor but keep it accessible from the code for backwards compatibility?

File: scene/3d/gpu_particles_collision_3d.cpp (hunk 1 above)
Comment (yahiaetman) [reply scene/3d/gpu_particles_collision_3d.cpp]: Fixed in https://github.com/godotengine/godot/pull/91858/commits/ebe4fb1585e947d5a0f5f6ecfdf7e74e1ede41ea

**Comment by AThousandShips** [comment]
To handle the code style issues I'd suggest fixing it locally instead of relying on the CI, see [here](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#using-clang-format-locally)

**Comment by yahiaetman** [comment]
I would like to open a discussion about which set of layers should be used for particle collisions. Currently, this pull request adds a new set of layers specifically for 3D particle collision, but I wonder if this is the best option. I think there are 3 options and these are the pros and cons of each in my opinion.

1. **3D Render Layers:** This was option used by the `cull_mask` property in both `GPUParticlesCollision3D` and `GPUParticlesAttractor3D`. It was supposed to be tested against the visual layer of the `GPUParticles3D` nodes.
   * **Pros:**
      * We don't need to add a new mask/layer property in `GPUParticles3D`.
   * **Cons:**
      * We need to access the visual layer from `RenderingServer` since it is not stored in the particle storage.
      * There are only 20 visual layers which are already used for other visual culling purposes, so it may become a limitation in large projects.
      * It is unintuitive to use layers that represent visual aspects of objects for physics related tasks.
 
2. **3D Physics Layers**: The `GPUParticles3D` class would have a `collision_mask` property that would be tested against a `collision_layer` property in both `GPUParticlesCollision3D` and `GPUParticlesAttractor3D`. We could also reverse them (put `collision_layer` in `GPUParticles3D` and the `collision_mask` in the collision and attractor nodes), but I think the first option is more intuitive.
   * **Pros:**
      * It is more intuitive to use physics layers for physics related tasks.
      * There are 32 available layers for physics, so it is less likely that it will be a limitation.
   * **Cons:**
      * We would be mixing between layers used for physics nodes and particle nodes.
  
3. **New 3D Particle Collision Layers**: This is similar to option 2 but instead of using 3D Physics layers for the layer/mask properties, new layers are used. This is the option implemented in the pull request.
   * **Pros:**
      * It is more intuitive to use separate layers for particle collisions.
      * It would leave 32 available layers just for particle collisions and won't compete with visual or physics layers.
   * **Cons:**
      * It would add a new set of layers just for particle collisions which includes adding a new page in the project settings for these layers and more code to maintain down the line.
     
Personally, I am more inclined towards either option 2 or 3. What do you think?
Also would it be better to put the `collision_mask` in `GPUParticles3D` and the `collision_layer` in `GPUParticlesCollision3D` & `GPUParticlesAttractor3D`, or vice versa?

**Comment by Calinou** [comment]
> Personally, I am more inclined towards either option 2 or 3. What do you think?

I think either option 2 or 3 is fine, but I'm inclined towards using option 2 at first since it's much less code. Most of the time, you want your particle collisions to behave in a way similar to other kinds of collision.

We can always try using option 2 for now, and switching to option 3 later on if the limitations prove to be too much.

> Also would it be better to put the collision_mask in GPUParticles3D and the collision_layer in GPUParticlesCollision3D & GPUParticlesAttractor3D, or vice versa?

I think collision mask properties should go in the *colliding* body (i.e. GPUParticles3D), not the *collider* (i.e. GPUParticlesCollision3D/Attractor3D). The collision mask should be interpreted as "what I collide with" after all.

**Comment by yahiaetman** [comment]
Superseded by:
- https://github.com/godotengine/godot/pull/93291

**Pull request closed without merging**